### DISCUSSION TITLE
Error: Custom SQL fields cannot contain sub-queries. When upgrading from 0.999 to 2.01. SQL in Filters

### DISCUSSION THREAD

**Discussion by xxcmp1977** [apache/superset#23574] [General]
Hi!

I upgraded my superset from old 0.999dev to 2.01 and Got an error
Error: Custom SQL fields cannot contain sub-queries
on my charts
It's caused by Filter in most of my charts with SQL subquery in "Custom SQL" tab
Something like
Filters - Add - CUSTOM SQL
"WHERE"

ID not in (select id from someothertable where somefield='somevalue')

And everything works well when removing this filter

Is it a bug of new version?
And should I consider using other features to achieve same results ?

**Comment by xxcmp1977** [comment]
Generally, the problem is that I want to have a chart of sales by day from my database. 
But I should not include sales to some group of clients (employes)
Scheme is like this:

table Orders:
orderId
clientId
orderdate
ordersum

table Clients:
clientId
clientName
clientGroupId

table ClientGroups
clientGroupId
clientGroupName

When I was on 0.999 version of superset I added table Orders as my Dataset
Visualization type: Table
Aggregate query mode with Dimensions on orderdate column and Sum of ordersum as Metrics
Then I added filter with custom sql value as

clientId not in (select clientid from clients where clientGroupId=1)

It worked perfect

I know I can use custom sql with joined tables as datasource now and filter on joined clientGroupId value
But I thought there is some another way achieving it
May be with new filters ?

**Comment by mpalmujoki** [comment]
We have encountered the same problem. 

Has this been examined or did you find a solution to enable it?

**Comment by xxcmp1977** [comment]
No
I had to rewrite datasets 'cause of this

**Comment by fmannhardt** [comment]
Encountered the same problem. The issue seems to be an overly aggressive detection of sub queries in filters. For example trying to extract epoch from a PostgreSQL interval causes this. My guess it is the "FROM" keyword. For example in something as:

` extract(epoch FROM (ts_end - ts_start)) * 1000 AS duration_ms`

I logged this as an issue here:
https://github.com/apache/superset/issues/25072

**Comment by kperveen** [comment]
Getting the same error when I use JOINS

**Comment by LeftMonster** [comment]
+1 same issue here

**Comment by zero-stroke** [comment]
I was trying to just add some columns to a table when I got this error, without any kind of aggregation or anything. I was able to get around it by clicking on the arrow next to the column name, and going to "Custom SQL", then just putting some single quotes around the column name. Like ```'column name'```

**Reply by torzsmokus** [reply]
I think this is in fact a _different bug_ (that columns with spaces are not handled correctly). Thanks a lot for your input: as a newcomer to superset, trying to set up my very first chart, I was totally baffled by the error message (and the workaround suggested by others that weakens security)

**Comment by nathan-gilbert** [comment]
I am experiencing this too after upgrading to 4.1.1, but @zero-stroke solution fixes it. It does seem to be a bug though and that existing text in these columns wouldn't be flagged as SQL just because it has `()` or `%` in it.

<img width="368" alt="image" src="https://github.com/user-attachments/assets/cba340f3-9bd4-4c06-ad73-aedbe24cea5c" />

That fixes it, but we didn't have to do that on every column before 4.1.1

**Reply by nathan-gilbert** [reply]
Really, need double quotes and not single quotes to get back to where we were before 4.1.1, fwiw

**Reply by yalefresne** [reply]
I met same issue in version 4.1.1 with `SUM` function and only ` seems to work.
![Capture d’écran 2025-01-16 à 11 59 39](https://github.com/user-attachments/assets/a3c765be-d523-4305-abfe-a79ec060f5da)

**Reply by torzsmokus** [reply]
In my case (data source: MS SQL Server), instead of quotes, square brackets were needed. Also, one single space worked fine for me, I just needed to tinker with the fields having more than one space.
![image](https://github.com/user-attachments/assets/42fa7f7f-73b2-4228-a84d-7697a29a4943)

**Comment by nktnet1** [comment]
>  Error: Custom SQL fields cannot contain sub-queries

A workaround for this is by enabling the `ALLOW_ADHOC_SUBQUERY` feature flag. For example, in your `superset_config.py`, ensure this is present:

```py
FEATURE_FLAGS = {
  "ALLOW_ADHOC_SUBQUERY": True
}
```

(**source:** superset github issue [#25072](https://github.com/apache/superset/issues/25072#issuecomment-1691836969)).

Note that this flag was disabled by default for security reason, so use it with caution. You can read more about this here:
- https://github.com/apache/superset/pull/19242
